src/sources/udp.rs

```diff
@@ -0,0 +1,220 @@
+                .map(move |(mut bytes, addr)| {
+                    // A packet has come.
+                    // UDP processes messages per packet, where messages are separated by newline.
+                    let mut events = Vec::new();
+                    while let Some(Some(line)) = new_line_decoder
+                        .decode_eof(&mut bytes)
+                        .map_err(|e| error!("{}", e))
+                        .ok()
+                    {
+                        // One message
+                        let mut event = Event::from(line);
+
+                        event
+                            .as_mut_log()
+                            .insert_implicit(host_key.clone().into(), addr.to_string().into());
+
+                        trace!(
+                            message = "Received one event.",
+                            event = field::debug(&event)
+                        );
+
+                        events.push(event);
+                    }
+                    iter_ok(events)
+                })
+                // Flatten messages from single packet
+                .flatten()
+                .map_err(|e: io::Error| error!("error reading datagram: {:?}", e))
+                .forward(out)
+                .map(|_| info!("finished sending"))
+        }),
+    )
+}
+
+#[cfg(test)]
+mod test {
+    use super::UdpConfig;
+    use crate::event;
+    use crate::test_util::next_addr;
+    use crate::topology::config::{GlobalOptions, SourceConfig};
+    use futures::{future::poll_fn, sync::mpsc, try_ready, Future, Stream};
+    use std::{
+        mem,
+        net::{SocketAddr, UdpSocket},
+        thread,
+        time::Duration,
+    };
+    use tokio::prelude::Async;
+
+    fn send_lines<'a>(addr: SocketAddr, lines: impl IntoIterator<Item = &'a str>) -> SocketAddr {
+        let bind = next_addr();
+
+        let socket = UdpSocket::bind(bind)
+            .map_err(|e| panic!("{:}", e))
+            .ok()
+            .unwrap();
+
+        for line in lines {
+            assert_eq!(
+                socket
+                    .send_to(line.as_bytes(), addr)
+                    .map_err(|e| panic!("{:}", e))
+                    .ok()
+                    .unwrap(),
+                line.as_bytes().len()
+            );
+            // Space things out slightly to try to avoid dropped packets
+            thread::sleep(Duration::from_millis(1));
+        }
+
+        // Give packets some time to flow through
+        thread::sleep(Duration::from_millis(10));
+        // Done
+
+        bind
+    }
+
+    fn collect_n<T>(mut rx: mpsc::Receiver<T>, n: usize) -> impl Future<Item = Vec<T>, Error = ()> {
+        let mut events = Vec::new();
+
+        poll_fn(move || {
+            while events.len() < n {
+                let e = try_ready!(rx.poll()).unwrap();
+                events.push(e);
+            }
+            Ok(Async::Ready(mem::replace(&mut events, Vec::new())))
+        })
+    }
```

Comment:
This looks nice! Feel free to add it to the crate's `test_utils` module if you think it's something that could be useful for other tests too.

Reply:
Great, I will.